django_pg_simple_hll/migrations/0001_initial.sql

```diff
@@ -0,0 +1,161 @@
+-- The state transition function
+-- it takes the current state, and the two input values
+-- hll_agg_state is the current running state, made up of bucket keys and hashes
+--  it is an array of n_buckets + 2
+-- input is the element we are considering
+-- hll_precision is the precision we are using for the approximation
+--  it use used to calculate the number of buckets
+create or replace function hll_bucket(
+    hll_agg_state int[],
+    input anyelement,
+    hll_precision int
+) returns int[]
+language plpgsql immutable strict parallel safe as $$
+declare
+    -- eg. for 512 buckets, hll_precision is 9 because 2**9 -> 512
+    n_buckets int := POW(2, hll_precision);
+    -- transform the input into a text and hash it
+    hashed_input int := hashtext(input::text);
+    -- keep all the significant bits, without the sign
+    bucket_hash int := hashed_input & ~(1 << 31);
+    -- bucket the hash into one of n buckets
+    bucket_key int := hashed_input & (n_buckets - 1);
+    -- pre fetch the hash for occupying the corresponding bucket for the input
+    -- we add 2 because
+    --  1. postgres arrays are 1-indexed
+    --  2. we have to account for an extra first element
+    current_hash int := hll_agg_state[bucket_key + 2];
+begin
+    -- we can only handle precision up to 31 or 2,147,483,648 buckets
+    -- because hashtext produces a hash of int size with 31 useable bits
+    if hll_precision < 4 or hll_precision > 31 then
+        raise exception 'invalid hll_precision: % - must be between 4 (16 buckets) and 31 (2,147,483,648 buckets) inclusive',
+            hll_precision;
+    end if;
+    -- postgres squeezes null elements in the array, we don't want that,
+    -- so we add a first and last element to keep a fixed size
+    if array_length(hll_agg_state, 1) < n_buckets + 2 then
+        hll_agg_state[1] := n_buckets;
+        hll_agg_state[n_buckets + 2] := n_buckets;
+    end if;
+
+    if current_hash is null or current_hash > bucket_hash then
+        hll_agg_state[bucket_key + 2] := bucket_hash;
+    end if;
+    return hll_agg_state;
+end $$;
+
+
+-- sfunc with default precision of 9
```

Comment:
Can you do this by just providing a default for the parameter in the above function definition?

Reply:
I've tried to do this, but can't manage it. Not sure if there is a trick

Reply:
I don't think this is possible. 

`CREATE AGGREGATE's` assumes it has the same singature than that of the `SFUNC`. So, we could setup a default argument to `hll_bucket`

```sql
CREATE OR REPLACE FUNCTION hll_bucket(
    hll_agg_state int [],
    input anyelement,
    hll_precision int DEFAULT 9
) RETURNS int []
...
```

but we cannot provide a default for the `AGGREGATE`.